File: CardMasterCard/Cards/CardsProject.vb
Imports System.IO
Imports System.Xml.Serialization
Imports Newtonsoft.Json


Namespace Card

    Public Class CardsProject

        Public Property Cards As CardsSet

        Public Shared Function LoadProject(file As FileInfo) As CardsProject

            Dim cardsProject As CardsProject = Nothing

            Select Case LCase(file.Extension)
                Case ".xml" : cardsProject = LoadXmlProject(file)
                Case ".json" : cardsProject = LoadJsonProject(file)
            End Select

            Return cardsProject

        End Function

        Private Shared Function LoadXmlProject(file As FileInfo) As CardsProject

            Dim CardsProject As CardsProject

            Dim myFileStream As Stream = file.OpenRead()
            Dim xs As New XmlSerializer(GetType(CardsProject))

            CardsProject = xs.Deserialize(myFileStream)

            myFileStream.Close()
            myFileStream.Dispose()
            myFileStream = Nothing

            Return CardsProject

        End Function

        Private Shared Function LoadJsonProject(file As FileInfo) As CardsProject

            Dim sr As New StreamReader(file.FullName)
            Dim js As String = sr.ReadToEnd

            Dim cardsProject As CardsProject = JsonConvert.DeserializeObject(Of CardsProject)(js)

            sr.Close()
            sr.Dispose()
            sr = Nothing

            Return cardsProject

        End Function

        Public Sub Save(ByRef file As FileInfo)

            Select Case LCase(file.Extension)
                Case ".xml" : SaveXmlProject(file)
                Case ".json" : SaveJsonProject(file)
            End Select

        End Sub

        Private Sub SaveXmlProject(ByRef file As FileInfo)

            Dim xs As New XmlSerializer(GetType(CardsProject))
            Dim xwr As New StreamWriter(file.FullName)

            xs.Serialize(xwr, Me)

            xwr.Close()
            xwr.Dispose()
            xwr = Nothing

        End Sub

        Private Sub SaveJsonProject(ByRef file As FileInfo)

            file.Delete()

            Dim js As String = JsonConvert.SerializeObject(Me, Formatting.Indented)
            Dim sw As New StreamWriter(file.FullName, FileMode.OpenOrCreate)

            sw.Write(js)

            sw.Close()
            sw.Dispose()
            sw = Nothing

        End Sub

    End Class

End Namespace

File: CardMasterCard/Cards/CardsProject.cs
using Newtonsoft.Json;
using System;
using System.IO;
using System.Xml.Serialization;

namespace CardMasterCard.Card
{
    public class CardsProject
    {
        public CardsSet Cards { get; set; }

        public static CardsProject LoadProject(FileInfo file)
        {
            CardsProject cardsProject = null;

            switch(file.Extension.ToLower())
            {
                case ".xml" :
                    cardsProject = LoadXmlProject(file);
                    break;
                case ".json" :
                    cardsProject = LoadJsonProject(file);
                    break;
            }

            return cardsProject;

        }

        private static CardsProject LoadXmlProject(FileInfo file)
        {
            CardsProject cardsProject = null;

            Stream myFileStream = file.OpenRead();
            XmlSerializer xs = new XmlSerializer(cardsProject.GetType());

            cardsProject = (CardsProject)xs.Deserialize(myFileStream);

            myFileStream.Close();
            myFileStream.Dispose();
            myFileStream = null;

            return cardsProject; 
        }

        private static CardsProject LoadJsonProject(FileInfo file)
        {
            StreamReader sr = new StreamReader(file.FullName);
            String js = sr.ReadToEnd();

            CardsProject cardsProject = (CardsProject)JsonConvert.DeserializeObject<CardsProject>(js);

            sr.Close();
            sr.Dispose();
            sr = null;

            return cardsProject;
        }

        public void Save(FileInfo file)
        {
            switch (file.Extension.ToLower())
            {
                case ".xml" :
                    SaveXmlProject(file);
                    break;
                case ".json" :
                    SaveJsonProject(file);
                    break;
            }
        }

        private void SaveXmlProject(FileInfo file)
        {
            XmlSerializer xs = new XmlSerializer(this.GetType());
            StreamWriter xwr = new StreamWriter(file.FullName);

            xs.Serialize(xwr, this);

            xwr.Close();
            xwr.Dispose();
            xwr = null;
        }

        private void SaveJsonProject(FileInfo file)
        {
            file.Delete();

            String js = JsonConvert.SerializeObject(this, Formatting.Indented);
            StreamWriter sw = new StreamWriter(file.FullName, false);

            sw.Write(js);

            sw.Close();
            sw.Dispose();
            sw = null;
        }

    }
}
